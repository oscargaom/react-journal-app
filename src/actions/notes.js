import { db } from "../firebase/firebase-config";
import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";
import { fileUpload } from "../helpers/fileUpload";

import Swal from "sweetalert2";

export const startNewNote = () => {
    /*  También podemos obtener el estado a través de una función que es enviada 
        por thunk como parámetro junto con el dispatch. Gracias a ello puedemos 
        tener acceso a todo el state y a los reducers y sería equivalente a usar
        el hook useSelector.
    */
    return async (dispatch, getState) => {
        // console.log(getState());
        const { auth: { uid } } = getState();
        // console.log(uid);

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        };

        const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

        dispatch(activeNote(doc.id, newNote));
    };
};

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});


export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        // console.log(notes);
        dispatch(setNotes(notes));
    };
};


export const startSaveNote = (note) => {
    return async (dispatch, getState) => {

        const { uid } = getState().auth;

        // console.log('startSaveNote note');
        // console.log(note);
        const noteToUpdate = { ...note };
        // console.log(noteToUpdate);
        delete noteToUpdate.id;
        // console.log(noteToUpdate);

        if (!noteToUpdate.url) {
            delete noteToUpdate.url;
        }

        try {
            await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToUpdate);
        } catch (error) {
            Swal.fire('Cant saved your note', error.toString(), 'error')
        }

        // console.log('startSaveNote note.id');
        // console.log(note.id);
        dispatch(refreshNote(note.id, note));

        Swal.fire('Saved note', 'Your note has been saved', 'success');
    };
};

export const setNotes = (notes) => {
    return {
        type: types.notesLoad,
        payload: notes
    };
};

export const refreshNote = (id, note) => {
    return {
        type: types.notesUpdated,
        payload: note
    };
};

export const startUploading = (file) => {
    return async (dispatch, getState) => {

        const { notes: { active } } = getState();

        // console.log(file);
        // console.log(active);

        Swal.fire({
            title: 'Uploading ...',
            text: 'Please wait ...',
            allowOutsideClick: false,
            onBeforeOpen: () => {
                Swal.showLoading();
            }
        });

        const fileUrl = await fileUpload(file);
        active.url = fileUrl;

        // console.log(fileUrl);

        dispatch(startSaveNote(active));

        Swal.close();
    }
}

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        const { auth: { uid } } = getState();

        // console.log(`startDeleting uid: ${uid}`);

        await db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch(deleteNote(id));
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})


export const cleanNotes = () => ({
    type: types.notesLogoutCleaning,
    payload: null
})