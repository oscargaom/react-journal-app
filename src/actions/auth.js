import { types } from "../types/types";
import Swal from 'sweetalert2'
import { firebase, googleAuthProvider } from '../firebase/firebase-config'
import { startLoading, finishLoading } from "./ui";
import { cleanNotes } from "./notes";

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                const { displayName, uid } = user;
                // console.log(displayName);
                // console.log(uid);
                dispatch(login(displayName, uid));
            });
    };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        // console.log(email);
        // console.log(password);
        // console.log(name);

        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async (userCredential) => {
                // console.log(userCredential);
                const { user } = userCredential;
                await user.updateProfile({ displayName: name });
                const { displayName, uid } = user;
                // console.log(displayName);
                // console.log(uid);

                // console.log(user);
                // console.log(displayName);
                Swal.fire('Bienvenido', 'Tu cuenta ha sido creada', 'success');
                dispatch(login(uid, displayName));

            })
            .catch(error => {
                // console.log(error);
                Swal.fire('Register error', error.message, 'error')
            });
    };
}

/*  Esta es una acción asíncrona, por ello en vez de regresar un objeto regresa 
    un callback el cual el middleware thunk queda en espera de la ejecución de algún
    dispatch; en caso de que no se ejecutará algún dispatch simplmente termina. 
    El dispatch es proporcionado por el middleware thunk, así es que no debemos 
    preocuparnos por obtenerlo de algún otro lugar.
*/
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        // setTimeout(() => {
        //     dispatch(login('Gpe', '456789'));
        // }, 3000);


        dispatch(startLoading());

        // setTimeout(() => {
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then((userCredential) => {
                dispatch(finishLoading());
                const { user: { displayName, uid } } = userCredential;
                Swal.fire('Bienvenido',`Hola ${displayName}`,'success');
                dispatch(login(uid, displayName));
            })
            .catch(error => {
                // const {code} = error;
                dispatch(finishLoading());
                dispatch(login('', ''));
                Swal.fire('Login Error', error.message, 'error');
            });
        // }, 9000);
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const logout = () => ({
    type: types.logout
});

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch(logout());

        dispatch(cleanNotes());
    };
}



