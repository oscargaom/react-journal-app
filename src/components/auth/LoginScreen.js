import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { useDispatch, useSelector } from 'react-redux';

import { startLoginEmailPassword, startGoogleLogin } from '../../actions/auth';

export const LoginScreen = () => {

    const [formValues, formHandleInputChange] = useForm({
        email: 'oscar@gmail.com',
        password: 'Hola123'
    });

    const { email, password } = formValues;

    const dispatch = useDispatch();

    const { ui: { loading } } = useSelector(state => state);

    // console.log(loading);

    const handleLogin = (e) => {
        e.preventDefault();
        // dispatch(login('987654', 'Manuel'));
        // console.log('dispatch startLoginEmailPassword');
        dispatch(startLoginEmailPassword(email, password));
    };

    const handleGoogleLogin = () => {
        // console.log('dispatch startGoogleLogin');
        dispatch(startGoogleLogin());
    };

    return (
        <>
            <h3 className="auth__title">Login</h3>
            <form 
                onSubmit={handleLogin}
                className="animate__animated animate__pulse animate__faster"
            >
                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={email}
                    onChange={formHandleInputChange}
                />

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={password}
                    onChange={formHandleInputChange}
                />

                <button
                    type="submit"
                    className="btn btn-primary btn-block"
                    disabled={loading}
                >
                    Login
                </button>

                <div className="auth__social-networks">
                    <p>Login with social networks</p>
                    <div
                        className="google-btn"
                        onClick={handleGoogleLogin}
                    >
                        <div className="google-icon-wrapper">
                            <img
                                alt="google button"
                                className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                            />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link
                    to="/auth/register"
                    className="link"
                >
                    Create new account
                </Link>
            </form>
        </>
    )
}
