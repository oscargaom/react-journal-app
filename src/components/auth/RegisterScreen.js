import React from 'react'
import validator from 'validator'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { setError, removeError } from '../../actions/ui'
import { useDispatch, useSelector } from 'react-redux'
import { startRegisterWithEmailPasswordName } from '../../actions/auth'

export const RegisterScreen = () => {

    const [formValues, formHandleInputChange] = useForm({
        name: 'oscar',
        email: 'oscar@gmail.com',
        password: 'Hola123',
        password2: 'Hola123'
    });

    const { name, email, password, password2 } = formValues;

    const dispatch = useDispatch();

    const state = useSelector(state => state);

    const { ui: { msgError } } = state;

    // console.log(msgError);

    const handleRegister = (e) => {

        e.preventDefault();

        if (isFormValid()) {
            // console.log('Formulario correcto');
            dispatch(startRegisterWithEmailPasswordName(email, password, name));
        }
    };

    const isFormValid = () => {

        if (name.trim().length === 0) {
            // console.log('Name is requiered');
            dispatch(setError('Name is requiered'));
            return false;
        } else if (!validator.isEmail(email)) {
            // console.log('Email is required');
            dispatch(setError('Email is required'));
            return false;
        } else if (password !== password2 || password.length < 5) {
            // console.log('Password should be at least 6 characters and match');
            dispatch(setError('Password should be at least 6 characters and match'));
            return false;
        }

        dispatch(removeError());

        return true;
    };

    return (
        <>
            <h3 className="auth__title">Register</h3>
            <form
                onSubmit={handleRegister}
                className="animate__animated animate__pulse animate__faster"
            >

                {
                    msgError !== null && (
                        <div className="auth__alert-error">
                            {msgError}
                        </div>
                    )
                }
                <input
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input"
                    autoComplete="off"
                    value={name}
                    onChange={formHandleInputChange}
                />

                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={email}
                    onChange={formHandleInputChange}
                />

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={password}
                    onChange={formHandleInputChange}
                />

                <input
                    type="password"
                    placeholder="Confirm password"
                    name="password2"
                    className="auth__input"
                    value={password2}
                    onChange={formHandleInputChange}
                />

                <button
                    type="submit"
                    className="btn btn-primary btn-block mb-5"
                >
                    Register
            </button>


                <Link
                    to="/auth/login"
                    className="link"
                >
                    Already register?
            </Link>
            </form>
        </>
    )
}
