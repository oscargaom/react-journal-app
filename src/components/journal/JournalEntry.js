import React from 'react'
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';

export const JournalEntry = ({ note }) => {

    // console.log(note);
        
    const {id, body, title, url, date} = note;
    
    const noteDate = moment(date);

    const dispatch = useDispatch();

    const handleEntryClick = () => {
        // console.log('handleEntryClick');
        dispatch(activeNote(id, note));
    };

    return (
        <div
            className="journal__entry pointer animate__animated animate__backInLeft animate__faster"
            onClick={handleEntryClick}
        >
            <div
                className="journal__entry-picture"
                style={{
                    backgroundSize: 'cover',
                    backgroundImage: `url(${url})`
                }}
            >
            </div>

            <div className="journal__entry-body">
                <p className="journal__entry-title">
                    {title}
                </p>
                <p className="journal__entry-content">
                    {body}
                </p>
            </div>

            <div className="journal__entry-date-box">
                <span>{noteDate.format('dddd')}</span>
                <h4>{noteDate.format('Do')}</h4>
            </div>
        </div>
    )
}
