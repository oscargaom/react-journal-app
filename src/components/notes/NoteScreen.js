import React, { useEffect, useRef } from 'react'
import { NotesAppBar } from './NotesAppBar'
import { useForm } from '../../hooks/useForm'
import { useSelector, useDispatch } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';

export const NoteScreen = () => {

    const dispatch = useDispatch();

    const { notes: { active: note } } = useSelector(state => state);

    // console.log(note);
    const [formValues, formHandleInputChange, reset] = useForm(note);

    const { body, title, url , id} = note;
    
    /*  useRef almacena una variable mutable que en este caso es note.id la 
        cual no va a redibujar a todo el componente NoteScreen si esta cambia.
    */
    const noteId = useRef(note.id);
    // console.log(noteId);
    
    useEffect(() => {
        /*  Como useForm utiliza su propio state debemos modificar el valor de
            dicho estado para que sea redibujado el componente con los valores
            de la nota, ya que de lo contrario siempre muestra el primer valor.
        */
        if (noteId.current !== note.id) {
            reset(note);
            noteId.current = note.id;
        }
    }, [reset, note]);

    
    useEffect(() => {
        // console.log(formValues);
        dispatch(activeNote(formValues.id, {...formValues}));
    }, [formValues, dispatch])
    
    const handleDelete = () => {
        // console.log(id);
        dispatch(startDeleting(id));
    }

    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">
                <input
                    type="text"
                    name="title"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    autoComplete="off"
                    value={title}
                    onChange={formHandleInputChange}
                />

                <textarea
                    name="body"
                    placeholder="What happened today"
                    className="notes__textarea"
                    value={body}
                    onChange={formHandleInputChange}
                >
                </textarea>

                {
                    (url)
                    &&
                    (<div className="notes__image">
                        <img
                            src={url}
                            alt="imagen"
                        />
                    </div>)
                }
            </div>

            <button
                className="btn btn-danger"
                onClick={handleDelete}
            >
                Delete
            </button>
        </div>
    )
}
