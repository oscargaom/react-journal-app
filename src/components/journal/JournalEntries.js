import React from 'react'
import { JournalEntry } from './JournalEntry';
import { useSelector } from 'react-redux';

export const JournalEntries = () => {

    //const entries =  [1,2,3,4,5];
    //const entries =  [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15];

    const stateSelector = useSelector(state => state);
    const { notes: entries } = stateSelector.notes;
    // console.log(entries);

    return (
        <div className="journal__entries">
            {
                entries && entries.map(note => (
                    <JournalEntry 
                        key={note.id} 
                        note={note}
                    />
                ))
            }
        </div>
    )
}
