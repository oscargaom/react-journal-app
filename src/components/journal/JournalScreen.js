import React from 'react'
import { Sidebar } from './Sidebar'
import { NoteScreen } from '../notes/NoteScreen'
import { NothingSelected } from './NothingSelected'
import { useSelector } from 'react-redux'

export const JournalScreen = () => {

    const stateSelector = useSelector(state => state);

    const { notes: { active } } = stateSelector;

    return (
        <div className="journal__main-content">

            <Sidebar />

            <main>
                {
                    (active)
                        ? <NoteScreen />
                        : <NothingSelected />
                }
            </main>
        </div>
    )
}
