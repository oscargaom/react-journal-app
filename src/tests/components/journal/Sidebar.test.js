import React from 'react'
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';

import { Sidebar } from '../../../components/journal/Sidebar';
import { Provider } from 'react-redux';
import { initialStateAuthNotes } from '../../utils/initialState';

import { startLogout } from '../../../actions/auth';
import { startNewNote } from '../../../actions/notes';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

jest.mock('../../../actions/auth', () => ({
    startLogout: jest.fn()
}));

jest.mock('../../../actions/notes', () => ({
    startNewNote: jest.fn()
}));

describe('Pruebas en el componente <Sidebar />', () => {

    let store = mockStore(initialStateAuthNotes);

    beforeEach(() => {
        store = mockStore(initialStateAuthNotes);
        /*  Al momento de ejecutarse dispatch(startLogout()) o cualquier otra acciÃ³n mediante dispatch 
            en las pruebas vamos a tener el siguiente error 'Actions must be plain objects. Use custom 
            middleware for async actions.' debido a que dispatch regresa una promesa y para evitar eso 
            hacemos uso de store.dispatch = jest.fn();
        */
        store.dispatch = jest.fn();
    });


    test('debe hacer match el snapshot.', () => {

        const wrapper = mount(
            <Provider store={store}>
                <Sidebar />
            </Provider>
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('debe llamar el dispatch startLogout mediante el evento handleLogout.', () => {

        const wrapper = mount(
            <Provider store={store}>
                <Sidebar />
            </Provider>
        );

        wrapper.find('button').prop('onClick')();

        expect(startLogout).toHaveBeenCalledTimes(1);
        expect(startLogout).toHaveBeenCalledWith();
    });

    test('debe de llamar el dispatch startNewNote mediante el evento handleNewEntry.', () => {
        const wrapper = mount(
            <Provider store={store}>
                <Sidebar />
            </Provider>
        );

        wrapper.find('.journal__new-entry').prop('onClick')();

        expect(startNewNote).toHaveBeenCalledTimes(1);
        expect(startNewNote).toHaveBeenCalledWith();
    })


})


