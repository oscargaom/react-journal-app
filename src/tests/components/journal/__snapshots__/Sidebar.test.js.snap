// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <Sidebar /> debe hacer match el snapshot. 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-moon"
      />
      <span>
         
        oscar
         
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
      New entry
    </p>
  </div>
  <div
    className="journal__entries"
  >
    <div
      className="journal__entry pointer animate__animated animate__backInLeft animate__faster"
      onClick={[Function]}
    >
      <div
        className="journal__entry-picture"
        style={
          Object {
            "backgroundImage": "url(https://res.cloudinary.com/dhc6rsfox/image/upload/v1597086578/p9npxjothlhhrdtkjasu.jpg)",
            "backgroundSize": "cover",
          }
        }
      />
      <div
        className="journal__entry-body"
      >
        <p
          className="journal__entry-title"
        >
          Titulo firebase
        </p>
        <p
          className="journal__entry-content"
        >
          trabajando duro!
        </p>
      </div>
      <div
        className="journal__entry-date-box"
      >
        <span>
          Monday
        </span>
        <h4>
          27th
        </h4>
      </div>
    </div>
    <div
      className="journal__entry pointer animate__animated animate__backInLeft animate__faster"
      onClick={[Function]}
    >
      <div
        className="journal__entry-picture"
        style={
          Object {
            "backgroundImage": "url(https://res.cloudinary.com/dhc6rsfox/image/upload/v1596500035/axoj3ozqkbfxxdch6tn1.jpg)",
            "backgroundSize": "cover",
          }
        }
      />
      <div
        className="journal__entry-body"
      >
        <p
          className="journal__entry-title"
        >
          Meditaci√≥n
        </p>
        <p
          className="journal__entry-content"
        >
          30 minutos.
        </p>
      </div>
      <div
        className="journal__entry-date-box"
      >
        <span>
          Monday
        </span>
        <h4>
          27th
        </h4>
      </div>
    </div>
  </div>
</aside>
`;
